File: android/app/src/main/java/com/budgetmanager/domain/interactor/AuthInteractor.kt
package com.budgetmanager.domain.interactor

import com.budgetmanager.domain.entity.AuthResponse
import com.budgetmanager.domain.entity.User
import com.budgetmanager.domain.repository.AuthRepository

class AuthInteractor(
    private val authRepository: AuthRepository
) {
    suspend fun register(email: String, password: String, name: String): Result<AuthResponse> {
        if (email.isBlank()) {
            return Result.failure(Exception("Email is required"))
        }
        if (!email.contains("@")) {
            return Result.failure(Exception("Invalid email format"))
        }
        if (password.length < 6) {
            return Result.failure(Exception("Password must be at least 6 characters"))
        }
        if (name.isBlank()) {
            return Result.failure(Exception("Name is required"))
        }

        return authRepository.register(email.trim(), password, name.trim())
    }

    suspend fun login(email: String, password: String): Result<AuthResponse> {
        if (email.isBlank()) {
            return Result.failure(Exception("Email is required"))
        }
        if (password.isBlank()) {
            return Result.failure(Exception("Password is required"))
        }

        return authRepository.login(email.trim(), password)
    }

    suspend fun logout() {
        authRepository.clearToken()
        authRepository.clearCurrentUser()
    }

    suspend fun isLoggedIn(): Boolean {
        return authRepository.getStoredToken() != null
    }

    suspend fun getCurrentUser(): User? {
        return authRepository.getCurrentUser()
    }

    suspend fun getToken(): String? {
        return authRepository.getStoredToken()
    }
}

File: iOS/Sources/Domain/Interactors/AuthInteractor.swift
import Foundation

@MainActor
class AuthInteractor: ObservableObject {
    private let authRepository: AuthRepositoryProtocol

    @Published private(set) var isLoggedIn: Bool = false
    @Published private(set) var currentUser: User?

    init(authRepository: AuthRepositoryProtocol) {
        self.authRepository = authRepository
        Task {
            await checkLoginStatus()
        }
    }

    func checkLoginStatus() async {
        let token = await authRepository.getStoredToken()
        isLoggedIn = token != nil
        if isLoggedIn {
            currentUser = await authRepository.getCurrentUser()
        }
    }

    func register(email: String, password: String, name: String) async throws -> AuthResponse {
        guard !email.trimmingCharacters(in: .whitespaces).isEmpty else {
            throw DomainError.invalidInput("Email is required")
        }

        guard email.contains("@") else {
            throw DomainError.invalidInput("Invalid email format")
        }

        guard password.count >= 6 else {
            throw DomainError.invalidInput("Password must be at least 6 characters")
        }

        guard !name.trimmingCharacters(in: .whitespaces).isEmpty else {
            throw DomainError.invalidInput("Name is required")
        }

        let response = try await authRepository.register(
            email: email.trimmingCharacters(in: .whitespaces),
            password: password,
            name: name.trimmingCharacters(in: .whitespaces)
        )

        isLoggedIn = true
        currentUser = response.user

        return response
    }

    func login(email: String, password: String) async throws -> AuthResponse {
        guard !email.trimmingCharacters(in: .whitespaces).isEmpty else {
            throw DomainError.invalidInput("Email is required")
        }

        guard !password.isEmpty else {
            throw DomainError.invalidInput("Password is required")
        }

        let response = try await authRepository.login(
            email: email.trimmingCharacters(in: .whitespaces),
            password: password
        )

        isLoggedIn = true
        currentUser = response.user

        return response
    }

    func logout() async {
        await authRepository.clearToken()
        await authRepository.clearCurrentUser()
        isLoggedIn = false
        currentUser = nil
    }
}
